File: reorder-list/Solution143.java
// https://leetcode.com/problems/reorder-list/

class Solution143 {

    // T: O(n), S: O(1)
    public void reorderList(ListNode head) {
        // Find the mid
        var slow = head;
        var fast = head.next;
        while (fast != null && fast.next != null) {
            slow = slow.next;
            fast = fast.next.next;
        }
        var first = head;
        var second = slow.next;
        // Sever the lists
        slow.next = null;
        // Reverse the second list
        ListNode prev = null;
        while (second != null) {
            var secondNext = second.next;
            second.next = prev;
            prev = second;
            second = secondNext;
        }
        second = prev;
        // Rebuild the list
        while (second != null) {
            var firstNext = first.next;
            var secondNext = second.next;
            first.next = second;
            second.next = firstNext;
            first = firstNext;
            second = secondNext;
        }
    }

    public static void main(String[] args) {
        var sol = new Solution143();
        var head1 = new ListNode(1, new ListNode(2, new ListNode(3, new ListNode(4))));
        System.out.println(head1); // 1 -> 2 -> 3 -> 4
        sol.reorderList(head1);
        System.out.println(head1); // 1 -> 4 -> 2 -> 3
    }

}

File: reorder-list/solution143.py
# https://leetcode.com/problems/reorder-list/

class ListNode:

    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

    def __repr__(self):
        result = []
        current = self
        while current:
            result.append(str(current.val))
            current = current.next
        return " -> ".join(result)


class Solution143:

    def reorderList(self, head):
        if not head or not head.next:
            return

        # Find the mid
        slow, fast = head, head.next
        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next

        first, second = head, slow.next
        # Sever the lists
        slow.next = None
        # Reverse the second list
        prev = None
        while second:
            second_next = second.next
            second.next = prev
            prev = second
            second = second_next
        second = prev
        # Rebuild the list
        while second:
            first_next = first.next
            second_next = second.next
            first.next = second
            second.next = first_next
            first = first_next
            second = second_next


def main() -> None:
    head1 = ListNode(1, ListNode(2, ListNode(3, ListNode(4))))
    print(head1)  # 1 -> 2 -> 3 -> 4
    sol = Solution143()
    sol.reorderList(head1)
    print(head1)  # 1 -> 4 -> 2 -> 3


if __name__ == "__main__":
    main()
